assayist/client/query.py

```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
     neomodel.db.set_connection(neo4j_url)
 
 
+def get_container_by_component(component_name, component_version):
+    """Query for builds by component name and version.
+
+    Finds builds that produced container images which include or embed a component with the
+    specified name and version, and any version preceding that version.
+
+    :param str component_name: name of the component to query for
+    :param str component_version: version of the component to query for
+    :return: list of builds IDs
+    :rtype: list
+    """
+    query = """
+        /// Find a component and a sourcelocation with the specified name and version. These can
+        /// be emdedded in other sourcelocation that are used as a source for a build or directly
+        /// used for a build.
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
+              (version_sl:SourceLocation {{canonical_version: '{version}'}})
+              <-[:UPSTREAM|EMBEDS*0..]- (target_sl:SourceLocation)
+              <-[:BUILT_FROM]- (direct_builds:Build)
+
+        /// Also, find all sourcelocation that precede the one with the specified version.
+        OPTIONAL MATCH (version_sl) -[:SUPERSEDES*]-> (:SourceLocation) <-[:UPSTREAM|EMBEDS*0..]-
+                       (older_target_sl:SourceLocation) <-[:BUILT_FROM]- (older_builds:Build)
+
+        /// Concatenate all found builds.
+        /// Note: we have to pass everything that we want to eventually use later in this WITH
+        /// statement otherwise it won't be available later on in the query.
+        WITH COLLECT(direct_builds) + COLLECT(older_builds) AS builds, target_sl, older_target_sl
+
+        /// Concatenate all found sourcelocations.
+        WITH COLLECT(target_sl) + COLLECT(older_target_sl) AS sls, builds
+
+        /// Expand lists into rows so we can match on them.
+        UNWIND builds as all_builds
+        UNWIND sls as all_sls
+
+        /// Find all container builds that embed an artifact produced by any of the previously
+        /// matched builds.
+        OPTIONAL MATCH (all_builds) -[:PRODUCED]-> (:Artifact) <-[:EMBEDS*0..]-
+                       (:Artifact) <-[:PRODUCED]- (container_build:Build {{type: 'container'}})
```

Comment:
I think nodes in all_builds can't also be in container_build. Not sure if that's a problem.

Reply:
Unless my thinking here is incorrect, that should be fine since this query is limited to returning container builds which will always embed the artifacts from builds that we are interested in. If this had to be generalized to RPM builds as well, `all_builds` would be a part of the results.